soco/music_services/music_service.py

```diff
@@ -0,0 +1,780 @@
+            soap_header=self.get_soap_header(),
+            namespace=self.namespace,
+            timeout=self.timeout)
+
+        try:
+            result_elt = message.call()
+        except SoapFault as exc:
+            if 'Client.TokenRefreshRequired' in exc.faultcode:
+                log.debug('Token refresh required. Trying again')
+
+                # <detail>
+                #   <refreshAuthTokenResult>
+                #       <authToken>xxxxxxx</authToken>
+                #       <privateKey>zzzzzz</privateKey>
+                #   </refreshAuthTokenResult>
+                # </detail>
+                auth_token = exc.detail.findtext('.//authToken')
+                private_key = exc.detail.findtext('.//privateKey')
+                # We have new details - update the account
+                self.music_service.account.oa_device_id = auth_token
+                self.music_service.account.key = private_key
+                message = SoapMessage(
+                    endpoint=self.endpoint,
+                    method=method,
+                    parameters=args,
+                    http_headers=self.http_headers,
+                    soap_action="http://www.sonos.com/Services/1"
+                                ".1#{0}".format(method),
+                    soap_header=self.get_soap_header(),
+                    namespace=self.namespace,
+                    timeout=self.timeout)
+                result_elt = message.call()
+
+            else:
+                raise MusicServiceException(exc.faultstring, exc.faultcode)
+
+        # The top key in the dict will be the methodResult. Its
+        # value may be None if no results were returned.
+        result = parse(
+            XML.tostring(result_elt), process_namespaces=True,
+            namespaces={'http://www.sonos.com/Services/1.1': None}
+        ).values()[0]
+
+        return result if result is not None else {}
+
+
+# pylint: disable=too-many-instance-attributes
+class MusicService(object):
+
+    """ The MusicService class provides access to third party music services.
+
+    Example:
+
+        Print all the services Sonos knows about
+
+        >>> from soco.music_services import MusicService
+        >>> print(MusicService.get_all_music_services_names())
+        ['Spotify', 'The Hype Machine', 'Saavn', 'Bandcamp',
+         'Stitcher SmartRadio', 'Concert Vault',
+         ...
+         ]
+
+        Or just those tho which you are subscribed
+
+        >>> print(MusicService.get_subscribed_services_names())
+        ['Spotify', 'radioPup', 'Spreaker']
+
+        Interact with TuneIn
+
+        >>> tunein = MusicService('TuneIn')
+        >>> print (tunein)
+        <MusicService 'TuneIn' at 0x10ad84e10>
+
+        Browse an item. By default, the root item is used. An OrderedDict is
+        returned.
+
+        >>> from json import dumps # Used for pretty printing ordereddicts
+        >>> print(dumps(tunein.get_metadata(), indent=4)
```

Comment:
Unfortunately, `pprint` does not work on OrderedDicts.  This is noted as a python bug. This seems to be generally accepted workaround for the moment